septa-fare-calculator/src/App.jsx

```diff
@@ -0,0 +1,212 @@
+// acceisibile form fields
+// form validation
+// styling
+// mobile styling?
+
+function App() {
+    const [zones, setZones] = useState([]);
+
+    // assumption:
+    // form is dynamicly calculating purchase costs
+    // with no submit button to get all form values at once
+    // best to access on form change form values via refs rather than local state to minimize renders
+    const formRef = useRef();
+    // defining current as an empty object prior to render
+    // allows me to use a single ref for all form fields
+    // 'current' is undefined on first render
+    // this is a problem when trying to set ref values:  ref => formRef.current.zones = ref
+    // defining here
+    formRef.current = {};
+    formRef.current.purchaseType = {};
+
+    // fetch fares
+    useEffect(() => {
+        const fetchFares = async () => {
+            const response = await fetch("./fares.json", {
+                headers: {
+                    "Content-Type": "application/json",
+                    Accept: "application/json",
+                },
+            }).then((res) => res.json());
+
+            setZones(response.zones);
+        };
+        fetchFares();
+    }, []);
+
+    useEffect(() => {
+        if (zones.length) {
+            const defaultFare = zones[0].fares[0];
+            calcFare(defaultFare);
+        }
+    }, [zones]);
+
+    // not sure how to manipulate radio buttons from a different form field update
+    // using this function to do so after 'anytime' travel is selected
+    const selectPurchaseType = (e) => {
+        const { selected } = formRef.current.purchaseType;
+        const tempUnselected = selected;
+        formRef.current.purchaseType.selected = e.target;
+        formRef.current.purchaseType.unselected = tempUnselected;
+    };
+
+    // not sure how to manipulate radio buttons from a different form field update
+    // using this function to do so after 'anytime' travel is selected
+    // and advanced purchase is not selected
+    const setAdvancedPurchaseType = (advancedPurchaseType) => {
+        const { selected, unselected } = formRef.current.purchaseType;
+        if (unselected.value === advancedPurchaseType) {
+            unselected.checked = true;
+            const tempUnselected = selected;
+            const tempSelected = unselected;
+            // using full object path to make the swap explicit to the reader
+            formRef.current.purchaseType.selected = tempSelected;
+            formRef.current.purchaseType.unselected = tempUnselected;
+        }
+    };
+
+    const calcFare = (fare) => {
+        const { purchaseCost, purchaseCount } = formRef.current;
+        // purchase value must be a minimum of 1
+        if (!purchaseCount.value) {
+            purchaseCost.innerText = MINIMUM_TICKET_PURCHASE;
+        } else {
+            purchaseCost.innerText =
+                "$" + parseInt(purchaseCount.value) * fare.price;
+        }
+    };
```

Comment:
The original implementation I had of this code ensured that a user must purchase at least 1 ticket
```
if (!purchaseCount.value) purchaseCount.value = 1
```

But I found, during testing, that this created an unpleasant experience where a user, for example, could not clear the field to write `2`.

The field defaults to 1. So, to write `2`, they had to tack on a `2` to `1`, creating `12` or `21` and then delete the '1' to use `2`